### PR TITLE
http core compiles (no tests yet)

### PR DESCRIPTION
**Pull request by drewhk**

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 1]
```diff
@@ -74,6 +74,14 @@ final class Flow[-In, +Out, +Mat](m: StreamLayout.Module, val inlet: Graphs.InPo
       inlet)
   }
 
+  def mapMaterialized[Mat2](f: Mat ⇒ Mat2): Repr[Out, Mat2] = {
+    val sourceCopy = module.carbonCopy()
+    new Flow(
+      sourceCopy.module.transformMaterializedValue(f.asInstanceOf[Any ⇒ Any]),
+      sourceCopy.inPorts(inlet).asInstanceOf[Graphs.InPort[In]],
+      sourceCopy.outPorts(outlet).asInstanceOf[Graphs.OutPort[Out]])
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: do we really need to copy all the ports here?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 2]
```diff
@@ -129,7 +136,7 @@ final class Flow[-In, +Out, +Mat](m: StreamLayout.Module, val inlet: Graphs.InPo
     source.via(this, (sm: Mat1, fm: Mat) ⇒ sm).to(sink, (sourcem: Mat1, sinkm: Mat2) ⇒ (sourcem, sinkm)).run()
   }
 
-  def section[I <: In, O, O2 >: Out, Mat2, Mat3](attributes: OperationAttributes, combine: (Mat, Mat2) ⇒ Mat3)(section: Flow[O2, O2, Unit] ⇒ Flow[O2, O, Mat2]): Flow[I, O, Mat3] = {
+  def section[O, O2 >: Out, Mat2, Mat3](attributes: OperationAttributes, combine: (Mat, Mat2) ⇒ Mat3)(section: Flow[O2, O2, Unit] ⇒ Flow[O2, O, Mat2]): Flow[In, O, Mat3] = {
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: I see you fixed the same bugs as I did :-)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 3]
```diff
@@ -65,6 +65,13 @@ final class Source[+Out, +Mat](m: StreamLayout.Module, val outlet: Graphs.OutPor
       .connect(outlet, sinkCopy.inlet))
   }
 
+  def mapMaterialized[Mat2](f: Mat ⇒ Mat2): Repr[Out, Mat2] = {
+    val sourceCopy = module.carbonCopy()
+    new Source(
+      sourceCopy.module.transformMaterializedValue(f.asInstanceOf[Any ⇒ Any]),
+      sourceCopy.outPorts(outlet).asInstanceOf[Graphs.OutPort[Out]])
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: same question here about the copying

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: you are right, we only need to copy when we import stuff

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: interestingly I added this to make it compile before, and now I removed it and it works :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: totally true, no need to copy here since we do not import anything

File: akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala [hunk 4]
```diff
@@ -89,34 +100,40 @@ private[http] object HttpClient {
         case (MessageStartError(_, info), _) ⇒ throw IllegalResponseException(info)
       }
 
-    import FlowGraphImplicits._
+    FlowGraph.partial { implicit b ⇒
+      import FlowGraph.Implicits._
+      val methodBypassFanout = Broadcast[HttpRequest](2)
+      val responseParsingMerge = b.add(new ResponseParsingMerge(rootParser))
```
Comment (bantonsson) [inline akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala]: This is not directly related to http, but I'm not super convinced that the implicit `FlowGraphBuilder` for the `Broadcast` and friends is not more confusing than convenient. My first reaction is: Why do you need to `add` the `ResponseParsingMerge` but not the `Broadcast`?

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala]: This is something that we want to unify. We discussed with roland to have a b.add() for the junctions as well. Shouldn't be hard, the question is if we want to make it mandatory and remove the sugared version that takes the implicit builder. Thils latter has the disadvantage that every test needs to be rewritten again.

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala]: Ah yes, so the idea was to have an "addToGraph" method that is basically the same as the current apply() but the name makes it more explicit what it does. Apply on the other hand will return the Graph, so it can be added with b.add()

Comment (rkuhn) [inline akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala]: incidentally that is exactly what I am working on right now

File: akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala [hunk 5]
```diff
@@ -159,108 +158,63 @@ private[http] object StreamUtils {
...
-              Some(result)
-            }
-          }
+      // FIXME FIXME FIXME: This is just to make the tests pass, this should not get into the real version
+      import scala.concurrent.Await
+      import scala.concurrent.duration._
 
-        case _ ⇒ None
-      }
+      import akka.actor.ActorSystem
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala]: Yes, this is a joke here. We need a proper solution, I just wanted to see if tests pass or not.

File: akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala [hunk 6]
```diff
@@ -159,108 +158,63 @@ private[http] object StreamUtils {
...
+            inputStream.close()
+            ByteString.empty
+          } else ByteString.fromArray(buffer, 0, read)
+        } else ByteString.empty
     }
 
-    new AtomicBoolean(false) with SimpleActorFlowSource[ByteString] {
-      override def attach(flowSubscriber: Subscriber[ByteString], materializer: ActorBasedFlowMaterializer, flowName: String): Unit =
-        create(materializer, flowName)._1.subscribe(flowSubscriber)
+    // FIXME FIXME FIXME: This should take the fileIODispatcher somehow, see commented out line
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala]: Unfortunately no such configuration is avaiable here. As you see the original implementation used a custom source. Probably I will take that route, too.

File: akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala [hunk 7]
```diff
@@ -132,24 +129,26 @@ private[http] object StreamUtils {
...
-
-            val broadcast = Broadcast[ByteString](OperationAttributes.name("transformMultipleInputBroadcast"))
-            input ~> broadcast
-            (multiple, results).zipped.foreach { (trans, sink) ⇒
-              broadcast ~> trans ~> sink
-            }
-          }.run()
-        results.map(s ⇒ Source(mat.get(s)))
-    }
+  def transformMultiple(input: Source[ByteString, Unit], transformers: immutable.Seq[Flow[ByteString, ByteString, _]])(implicit materializer: FlowMaterializer): immutable.Seq[Source[ByteString, Unit]] = ???
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala]: This is simply impossible with the current Streams DSL. I have no idea what to do here.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 8]
```diff
@@ -74,6 +74,14 @@ final class Flow[-In, +Out, +Mat](m: StreamLayout.Module, val inlet: Graphs.InPo
       inlet)
   }
 
+  def mapMaterialized[Mat2](f: Mat ⇒ Mat2): Repr[Out, Mat2] = {
```
Comment (bantonsson) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: The return of the materialized value transformer ;)

File: akka-http-core/src/main/java/akka/http/model/japi/HttpEntities.java [hunk 9]
```diff
@@ -42,19 +42,19 @@ public static UniversalEntity create(ContentType contentType, File file) {
...
         return new akka.http.model.HttpEntity.CloseDelimited((akka.http.model.ContentType) contentType, data);
     }
 
-    public static HttpEntityIndefiniteLength createIndefiniteLength(ContentType contentType, Source<ByteString> data) {
+    public static HttpEntityIndefiniteLength createIndefiniteLength(ContentType contentType, Source<ByteString, scala.runtime.BoxedUnit> data) {
         return new akka.http.model.HttpEntity.IndefiniteLength((akka.http.model.ContentType) contentType, data);
     }
 
-    public static HttpEntityChunked createChunked(ContentType contentType, Source<ByteString> data) {
+    public static HttpEntityChunked createChunked(ContentType contentType, Source<ByteString, scala.runtime.BoxedUnit> data) {
```
Comment (rkuhn) [inline akka-http-core/src/main/java/akka/http/model/japi/HttpEntities.java]: let’s do this in style and use an import ;-) (I know you thought this was a hack, but I think it is the only route we can take)

File: akka-http-core/src/main/java/akka/http/model/japi/HttpEntity.java [hunk 10]
```diff
@@ -73,5 +73,5 @@
     /**
      * Returns a stream of data bytes this entity consists of.
      */
-    public abstract Source<ByteString> getDataBytes();
+    public abstract Source<ByteString, scala.Unit> getDataBytes();
```
Comment (rkuhn) [inline akka-http-core/src/main/java/akka/http/model/japi/HttpEntity.java]: This is likely wrong.
